hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java

```diff
@@ -370,6 +384,26 @@ private void updateContainerReplica(final DatanodeDetails datanodeDetails,
     }
   }
 
+  /**
+   * Determines whether container replica is empty or not.
+   *
+   * @param replicaProto container replica details.
+   * @return true if container replica is empty else false
+   */
+  private boolean fillEmpty(final ContainerReplicaProto replicaProto) {
+    if (replicaProto.hasIsEmpty()) {
+      return replicaProto.getIsEmpty();
+    } else {
+      // Handled when DN version is old then there will not be isEmpty field.
+      // In this case judge container empty based on keyCount
+      // and bytesUsed field.
+      if (replicaProto.getKeyCount() == 0 && replicaProto.getUsed() == 0) {
+        return true;
+      }
```

Comment:
This value should be populated on DN startup when it is loading the containers into memory. There should not be a case where this value is unset since we do not support rolling upgrades with mixed versions yet.

Reply:
What about when we do support rolling upgrades? Someone may be on a version where the DNs don't have the new field and then jump to a future version that does support rolling upgrades. IMO it is safer to try to be compatible, rather than saying "we don't support that" as that support may change. There have been quite a few places in the code where I have seen this sort of compatibility code, and I think it starts a dangerous precedent if we are not saying we don't need to do it.

Reply:
@errose28 @sodonnel, I have kept this change for now. So can we go ahead we this?

Reply:
Compatibility guarantees have to be strictly defined and tested. For example, we support non-rolling upgrades from all versions since 1.0.0 and non-roling downgrades pre-finalization to all versions since 1.1.0. We have tests for this and require changes to adhere to these rules. The compatibility code you are referring to is to comply to these guidelines.

Rolling upgrade has not been planned yet. When it is, we will define an earliest version that supports rolling upgrade, tests, and requirements that developers must follow, just like we have for non-rolling upgrade. Without this, enforcement of rolling upgrade compatibility will be arbitrary. We cannot attempt to support the feature in this state. There are also issues that guarantee no code getting checked in to master right now will be eligible for rolling upgrade support:

- OM HA commit model is incompatible with rolling upgrades. There may be other areas of the code as well that currently prevent us from doing rolling upgrades from this version, but we have not done an invovled investigation yet.

- There are no test suites to catch any rolling upgrade incompatibilities that are introduced, so it is impossible to guarantee the current version would work with a rolling upgrade. If it happend to work it would be purely by coincidence.

- There is no agreement among developers on rolling upgrade version guarantees so requirements are not being enforced in code reviews.

Once the rolling upgrade feature is proposed and requirements are defined then we will have explicit rules to guide changes and versions. Without this framework in place, enforcing rolling upgrade requirements in PRs just leads to confusing code, because it is documenting and designing for a situation that cannot happen. For that reason I propose removing the extra check, even though it does not affect correctness in the current version.

Reply:
Since rolling upgrade is not yet supported, Once it starts supporting that will be first version and future code needs to handle accordingly. So before support there will not be recommendation to do rolling upgrade. 
@sodonnel @errose28, Please confirm shall I remove this compatibility change?

Reply:
OK - it is fine to remove the compatibility part. I am sure I have seen (and had review comments against my own code) code in the past that does this compatibility thing, which I why I raised it, but if we really don't need to worry about it for now, then that is good to know for the future.